### PR TITLE
Compile error when deferred mapper was not created

### LINKED ISSUE #1904: Mappers using @Context not yielding Impl classes in Java11
While transitioning from Java 8 to Java 11 I found that any mappers that were using the `@Context` annotation stopped generating Impl classes, leading to "Cannot find implementation" errors at startup or test time. I've been able to reproduce using the CarDto mapper outlined in the reference guide:

```java
@Mapper
public interface CarMapper {
    default CarManualDto translateManual(CarManual manual, @Context Locale locale) {
        return new CarManualDto(manual.getContent());
    }

    CarDto toCar(Car car, @Context Locale locale);
}
```

When compiled using Java 8, we get the following Impl.java:
```java
@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2019-09-13T13:02:01-0500",
    comments = "version: 1.3.0.Final, compiler: javac, environment: Java 1.8.0_121 (Oracle Corporation)"
)
public class CarMapperImpl implements CarMapper {

    @Override
    public CarDto toCar(Car car, Locale locale) {
        if ( car == null ) {
            return null;
        }

        CarDto carDto = new CarDto();

        carDto.setManual( translateManual( car.getManual(), locale ) );

        return carDto;
    }
}
```

When compiled using Java 11+ (with a source and target of 1.8), this Impl.java file is no longer being generated (using the same minimal maven pom):

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>org.mapstruct</groupId>
  <artifactId>context-bug</artifactId>
  <version>1.0-SNAPSHOT</version>

  <dependencies>
    <dependency>
      <groupId>org.mapstruct</groupId>
      <artifactId>mapstruct</artifactId>
      <version>1.3.0.Final</version>
    </dependency>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.8.0</version>
        <configuration>
          <source>1.8</source>
          <target>1.8</target>
          <annotationProcessorPaths>
            <path>
              <groupId>org.mapstruct</groupId>
              <artifactId>mapstruct-processor</artifactId>
              <version>1.3.0.Final</version>
            </path>
          </annotationProcessorPaths>
        </configuration>
      </plugin>
    </plugins>
  </build>
</project
```

Note that this issue *may* have started earlier than Java 11, but I've only tested with Java 11 and Java 12.

### PR DESCRIPTION
**Pull request by filiphr**
Additionally exclude private methods when fetching enclosed executable elements

Fixes #1904

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java [hunk 1]
```diff
@@ -85,8 +85,8 @@ private static TypeElement asTypeElement(TypeMirror mirror) {
 
     /**
      * Finds all executable elements within the given type element, including executable elements defined in super
-     * classes and implemented interfaces. Methods defined in {@link java.lang.Object} are ignored, as well as
-     * implementations of {@link java.lang.Object#equals(Object)}.
+     * classes and implemented interfaces. Methods defined in {@link java.lang.Object},
+     * implementations of {@link java.lang.Object#equals(Object)} and private methods are ignroed
```
Comment (Captain1653) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java]: small typo "ignored" :)

File: processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java [hunk 2]
```diff
@@ -168,6 +168,15 @@ private static boolean isNotObjectEquals(ExecutableElement executable) {
         return true;
     }
 
+    /**
+     * @param executable the executable to check
+     *
+     * @return {@code true}, iff the executable does not have a private modifier
```
Comment (Captain1653) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java]: small typo "if"

File: processor/src/main/java/org/mapstruct/ap/MappingProcessor.java [hunk 3]
```diff
@@ -347,4 +378,15 @@ public int compare(ModelElementProcessor<?, ?> o1, ModelElementProcessor<?, ?> o
             return Integer.compare( o1.getPriority(), o2.getPriority() );
         }
     }
+
+    private static class DeferredMapper {
```
Comment (Captain1653) [inline processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: I'm not sure. Should we write some javadoc about the purpose of this private static class? Or is it obvious?

File: processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java (hunk 2 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java]: iff is shortened for if and only if 😄

File: processor/src/main/java/org/mapstruct/ap/MappingProcessor.java (hunk 3 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: I thought that it was obvious from the context. If it isn't I can add some javadoc

Comment (Captain1653) [reply processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: No, you are right

File: processor/src/main/java/org/mapstruct/ap/MappingProcessor.java [hunk 4]
```diff
@@ -163,6 +164,32 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
...
+                String erroneousElementName;
+
+                if ( erroneousElement instanceof QualifiedNameable ) {
+                    erroneousElementName = ( (QualifiedNameable) erroneousElement ).getQualifiedName().toString();
+                }
+                else {
+                    erroneousElementName = erroneousElement.getSimpleName().toString();
+                }
+
+                processingEnv.getMessager()
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: is there any way to figure out what annotatiotion processors are in the compilation cycle? If so, it we should list them. If not: it seems that the error message needs to point more clearly to the underlying cause in a more generic way. something along the lines what you have now, but a hint that another ap (not generating classes) might be the problem.

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: I don't think that we get can the current annotation processors. But if you have some better idea for the message I would gladly hear it. 

I can add the hint that maybe some another annotation processor didn't create the class and that you could enable verbose mode for mapstruct so you can see what is happening.

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: I changed the text a bit let me know what you think

File: processor/src/main/java/org/mapstruct/ap/MappingProcessor.java [hunk 5]
```diff
@@ -163,6 +164,40 @@ public boolean process(final Set<? extends TypeElement> annotations, final Round
...
+                // Otherwise the reporting will not work properly
+                deferredMapperElement = annotationProcessorContext.getElementUtils()
+                    .getTypeElement( deferredMapperElement.getQualifiedName() );
+
+                processingEnv.getMessager()
+                    .printMessage(
+                        Kind.ERROR,
+                        "No implementation was created for " + deferredMapperElement.getSimpleName() +
+                            " due to having a problem in the erroneous element " + erroneousElementName + "." +
+                            " Hint: this usually means that some other annotation processor was supposed to" +
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/MappingProcessor.java]: usually == often?

**Comment by filiphr** [comment]
Thanks for the review @sjaakd and @Captain1653

**Pull request merged**